packages/SwingSet/src/kernel/vatTranslator.js

```diff
@@ -122,23 +144,39 @@ function makeTranslateKernelDeliveryToVatDelivery(vatID, kernelKeeper) {
     return reapMessageVatDelivery;
   }
 
+  /**
+   *
+   * @param { KernelDeliveryObject } kd
+   * @returns { VatDeliveryObject }
+   */
   function kernelDeliveryToVatDelivery(kd) {
-    const [type, ...args] = kd;
-    switch (type) {
-      case 'message':
+    switch (kd[0]) {
+      case 'message': {
```

Comment:
I like the "before" code that destructures into named consts much better than the "after" code that directly accesses `kd` as an array.  It does a superior job of communicating intent.  I advocate reverting this change.

Reply:
I hate it too, but it was the only way @michaelfig and I could find to appease the typechecker. The core issue (if I understand properly) is that typescript isn't clever enough to combine `const [a,b] = c;` and a subsequent constraint on `a` into a deduced constraint on `c`. So even though the `case 'message'` tells us that `type` must be `message`, and therefore `kd` must be a `KernelDeliveryMessage`, the typechecker doesn't realize that. But `kd[0] === 'message'` *is* enough to provoke that deduction.

The same sort of issue is what requires the second part: once it knows that `kd` is a `KernelDeliveryMessage` (within this one `case`), it knows what `kd[1]` is, and `const [_, ...args] = kd` allows it to know what `args` is. But if the factored out `[type, args] = kd` is before the switch, it *can't* deduce a type for `args`, and then all the calls to the more specific functions look like they're using the wrong (or an unknown) type.

This is clearly a tension between human readability and typechecker readability. I guess the deciding question is whether appeasing the typechecker gives us enough advantage elsewhere to make up for the impact to human readability. In this case, the code is ugly, but not unreadably ugly, so I think it's worth it.

But I'm really hoping that the typechecker gets smarter some day, so we can go back to the old code, because I agree, this looks gross.

Reply:
This is something that TypeScript can do in "native" TypeScript code, so I'm not convinced we can't make it work here.  I'm inclined to whine louder about this.  If TypeScript is forcing us to write obviously worse code just to satisfy the type checker then something is very wrong.  If we are going to be pushed into something that's harder to read, I would say that if a type annotation comment above the `const [a, b] = c; ` line could do the trick than I could live with that.  But this is just horrible.

Reply:
It makes sense that the earlier case didn't work because it couldn't know the length of `args` before the switch. Putting it under the case lets it constrain that.

Though it would be more more readable to ditch the _rest_,
```
 function kernelDeliveryToVatDelivery(kd) {
    switch (kd[0]) {
      case 'message': {
        return translateMessage(kd[1], kd[2]);
```


[playground example](https://www.typescriptlang.org/play?jsx=0&ts=4.2.3&ssl=29&ssc=1&pln=29&pc=2#code/C4TwDgpgBAwghmAInYcA8BlANge2APigF4oBvAWACgoaoAjHAExAC4oBnYAJwEsA7AOYAaKrQ65g7NtjwBtALojKAXwDcVUJCgYA7vwEYIweEhRxisBMlRpOvQfg3hoAWQjt2cAdBIVqAWyMACyY2O30lKDguASltPUFDYyszSK53AFcsYAB+MO4IqjUnLQBpCC4+CCxEap4ANwqQNw8vHyhZVAE2AHJA1u8eoShUGKN8+2Eof3ZuqBbPb3l1Sk1ocsrq2qwGpoA5PB4AMxALTq9evkOToah09hwsDOAeHD448MEFKGWqEvWKlUanVGlwQAB5OgAKwgAGNgBYNkDtrswQs2lAAD5QJFbEH7a4gFZURhwrDRaBHDJ8eGvPgjLhwd7k4AQdHeAAUo28wAmEWmszY7IgAEo2EyiSSyRSoFSaS83gymewWRADi8Thz7o9nnSPgUvvIxVE+JLKFRYW9OFAANaAvE7UEgAAqOAAaigUU6LBybYw2LjgY6mpCYfCRcRCH5RLR2HpgLCglBfYxZAAGI1kGNiWFwdjQPruRYQHosbNiO5GDKVJXMlBsottFOyACMiltqYATEaVhWoLn81AelcNSBS+WxOlgNX6dxlar1ccQM22yLexXSUc4FleRPaMAglwcDooFUTwBRLhHrgcgAG1JtVx09LtmyDqJAADo1lAACSkP103kZRbzXctlCKIA) (building off @mhofman's)

Reply:
> This is something that TypeScript can do in "native" TypeScript code

You mean in `.ts` files? What does it look like there?

Reply:
I don't recall, as I wasn't the one who did the TypeScript voodoo.  We had a guy who was a savant w.r.t H-M type systems.  I just remember we had a bunch of this kind of stuff.